### DISCUSSION TITLE
Am I caught in a push-pull catch-22?

### DISCUSSION THREAD

**Discussion by mhuzzell** [IBAHCM/RPiR#536] [Q&A for technical issues] [answered]
I pushed a draft of Practical 2-1 onto my repository at the end of class on Monday, intending to edit it and push the edit. It took a long time for the initial push to commit, which seemed to be because I had accidentally included the html report along with the R code. So, I went into the repository and manually deleted the html file.

I then went and finished Practical 2-1, hit commit, and then tried to push it. Which gives me this error message:

> 
> >>> C:/Program Files/Git/bin/git.exe push origin HEAD:refs/heads/main
> To https://github.com/SBOHVM/mhuzzellSeries02.git
>  ! [rejected]        HEAD -> main (non-fast-forward)
> error: failed to push some refs to 'https://github.com/SBOHVM/mhuzzellSeries02.git'
> hint: Updates were rejected because the tip of your current branch is behind
> hint: its remote counterpart. Integrate the remote changes (e.g.
> hint: 'git pull ...') before pushing again.
> hint: See the 'Note about fast-forwards' in 'git push --help' for details.

However, if I do try to pull first, it quite sensibly warns me that my edits to finish the practical would be overwritten, and aborts the merge:

> >>> C:/Program Files/Git/bin/git.exe pull
> error: Your local changes to the following files would be overwritten by merge:
> 	0201-run-SIS.html
> Please commit your changes or stash them before you merge.
> Aborting
> Merge with strategy ort failed.

I would like to "See the 'Note about fast-forwards' in 'git push --help' for details." but I'm not sure which 'help' repository to look in. Can any humans help me, please?

**Comment by mhuzzell** [comment]
Wait, just remembered I _did_ manage to edit my code in a more basic way earlier (just as a test), and that pushed fine. So now it is even more mysterious!

**Comment by mhuzzell** [comment]
Wait wait a second time, I went back and checked, and I deleted the apparently rogue html file _after_ pushing those first superficial edits. So I expect that deletion _is_ the problem. But how can I get out of it?

**Comment by richardreeve** [comment]
@mhuzzell Sorry for the delay in getting back to you. This is a classic problem with git, and for future reference really never, ever edit your code on GitHub (it sounds like this is where you first deleted the file). It causes all sorts of problems that even experienced git users try really hard not to get involved with.

However, it may not be that hard to fix... in person it'll almost definitely be easy, but here's something to try anyway. You should either see the html file in the git tab with either a `M` or a `D` symbol, like I've done here to a couple of files:

<img width="888" alt="Screenshot 2022-12-01 at 19 19 17" src="https://user-images.githubusercontent.com/3344796/205141222-55aaeff1-831b-4ba5-811b-7f531d3a7e63.png">

You may see some other files with `?` or other edits (`M`) that you have deliberately changed, but ignore them. Select and tick the html one that you want to remove, and then click `More`, and `Revert...`:

<img width="888" alt="Screenshot 2022-12-01 at 19 19 49" src="https://user-images.githubusercontent.com/3344796/205141600-ff8aaa2a-31f3-4498-8afa-4a3f92e44ab2.png">

and then yes. The file should disappear from the git tab, to say it has gone back to how it was before. You should now be able to "pull" the deletion from GitHub, and the file should disappear, and then you'll be able to commit and push your changes back... try that and see how it goes anyway...

**Comment by mhuzzell** [comment, marked as answer]
Resolved in class -- thank you for your help!
